core/src/main/java/org/infinispan/cache/impl/CacheImpl.java

```diff
@@ -499,7 +500,9 @@ public final boolean isEmpty() {
    }
 
    final boolean isEmpty(long explicitFlags) {
-      return !entrySet(explicitFlags).stream().anyMatch(StreamMarshalling.alwaysTruePredicate());
+      try (Stream<K> stream = keySet(explicitFlags).stream()) {
```

Comment:
These shouldn't be needed. A stream should be closed automatically if it is fully consumed, which `anyMatch` would do. This would be a sign of a different issue, which I am guessing you fixed with the `PersistenceEntryStreamSupplier` changes.

Reply:
Calls to isEmpty() hang the server on shutdown without this in place - so perhaps the issue is that this:

> A stream should be closed automatically if it is fully consumed

... isn't actually happening.

Reply:
The alternative to an explicit close() here would be to add explicit close() invocations to all of the terminal operations of AbstractLocalCacheStream.

Reply:
I was more thinking in the `SingleRunOperation` and `SegmentRetryingOperation`, just two places then.

Reply:
But I am not certain yet though either as it may have to be elsewhere. As I mentioned I am looking into it now.

Reply:
Ah - these are currently only used only for DistributedCacheStreams, not LocalCacheStreams.

Reply:
Yeah, as I mentioned I am looking into it :)

But I agree this needs to be done to all the implementors of `AbstractLocalCacheStream` in their method invocations. Technically only early termination operations should be affected, but there is no real reason to not do it to all terminal operations.

Reply:
On a side note https://issues.jboss.org/browse/ISPN-9813 will eventually replace all this on master :)

Reply:
> can you check my configuration and errors with transactional caches (see attached file TRANSACTIONAL.txt) to verify they match the issue in ISPN-10114?

That what i did initially, but given how many lines of code this involved (including the Double/Int stream derivations) I thought it might be preferable to address the specific Cache methods that looked problematic.

Anyway, would you prefer to take this over?  Or shall I update the PR?

Reply:
I can take it now, thanks. I should have a PR here in just hopefully an hour.

Reply:
> That what i did initially, but given how many lines of code this involved (including the Double/Int stream derivations) I thought it might be preferable to address the specific Cache methods that looked problematic.

Unfortunately these streams are available to the user to invoke directly, so they need to be plugged directly, not only at the Cache level. :(